### PR TITLE
[Routing] Allow to set name prefixes from the configuration

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #19612
| License       | MIT
| Doc PR        | ø

This allows setting name prefixes to routes while importing them. Typically, we can then import multiple times a similar file. This was originally requested by 🎸 @chrisguitarguy in https://github.com/symfony/symfony/issues/19612

```yaml
app:
    resource: ../controller/routing.yml

api:
    resource: ../controller/routing.yml
    name_prefix: api_
    prefix: /api
```

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<routes xmlns="http://symfony.com/schema/routing"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://symfony.com/schema/routing
        http://symfony.com/schema/routing/routing-1.0.xsd">

    <import resource="../controller/routing.xml" />
    <import resource="../controller/routing.xml" prefix="/api" name-prefix="api_" />

</routes>
```

### CODE REVIEW

File: src/Symfony/Component/Routing/Tests/RouteCollectionTest.php [hunk 1]
```diff
@@ -302,4 +302,17 @@ public function testSetMethods()
         $this->assertEquals(array('PUT'), $routea->getMethods());
         $this->assertEquals(array('PUT'), $routeb->getMethods());
     }
+
+    public function testAddNamePrefix()
+    {
+        $collection = new RouteCollection();
+        $collection->add('foo', $foo = new Route('/foo'));
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/RouteCollectionTest.php]: please also add an existing `api_foo` route, to be sure that the prefixing does not delete this route due to overriding.

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
Does the PHP DSL also need this feature, or is it already available in it ?

File: src/Symfony/Component/Routing/Tests/RouteCollectionTest.php (hunk 1 above)
Comment (sroze) [reply src/Symfony/Component/Routing/Tests/RouteCollectionTest.php]: Good catch, added.

**Comment by sroze** [comment]
@stof it was't. I've added it to the trait.

**Comment by sroze** [comment]
Status: Needs review

File: src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php [hunk 2]
```diff
@@ -124,4 +124,18 @@
 
         return $this;
     }
+
+    /**
+     * Add a prefix to the name of all the routes within the collection.
```
Comment (fabpot) [inline src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: `Adds`

File: src/Symfony/Component/Routing/RouteCollection.php [hunk 3]
```diff
@@ -153,6 +153,22 @@ public function addPrefix($prefix, array $defaults = array(), array $requirement
         }
     }
 
+    /**
+     * Add a prefix to the name of all the routes within in the collection.
```
Comment (fabpot) [inline src/Symfony/Component/Routing/RouteCollection.php]: `Adds`

File: src/Symfony/Component/Routing/RouteCollection.php [hunk 4]
```diff
@@ -153,6 +153,22 @@ public function addPrefix($prefix, array $defaults = array(), array $requirement
...
+     *
+     * @param string $prefix
```
Comment (fabpot) [inline src/Symfony/Component/Routing/RouteCollection.php]: Should be removed

File: src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php [hunk 5]
```diff
@@ -124,4 +124,18 @@
...
+     *
+     * @param string $prefix
```
Comment (fabpot) [inline src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: Should be removed

File: src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php [hunk 6]
```diff
@@ -124,4 +124,18 @@
...
+     *
+     * @return $this
+     */
+    final public function addNamePrefix(string $prefix)
```
Comment (chalasr) [inline src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: `: self` to fully get rid of the phpdoc?

Comment (sroze) [reply src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: Well, we still need the description I'd say, as it's meant to be used by developers

**Review by sroze — COMMENTED** [review]
Updated based on @fabpot's comments.

File: src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php (hunk 6 above)
Comment (fabpot) [reply src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: Getting rid of `@return $this` would be nice though.

Comment (sroze) [reply src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: but... on the trait I'm not sure if this makes sense... as (at least in my mind) `self` refers to the trait here, not the class `use`ing the trait.

Comment (fabpot) [reply src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: There is no such thing as referring to a trait. PHP merges traits into the class, so self will always point to the correct class.

Comment (sroze) [reply src/Symfony/Component/Routing/Loader/Configurator/Traits/RouteTrait.php]: I was talking conceptually, by reading this. But if you two believe it's better just to put `self` within the trait, I'm happy with it as I don't mind that much tbh. Changed 👍

**Review by Simperfit — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @sroze.

**Pull request closed without merging**